File: solutions/2026-04-06/java/2026-04-06-walking-robot-simulation.java
```java
import java.util.HashSet;
import java.util.Set;

// Problem: Walking Robot Simulation
// Link: https://leetcode.com/problems/walking-robot-simulation/
//
// Approach:
// We simulate the robot's movement step by step.
// We maintain the robot's current coordinates (x, y) and its current direction.
// The directions can be represented by an integer: 0 for North (+y), 1 for East (+x), 2 for South (-y), 3 for West (-x).
// When the robot turns left (-2), we update the direction: (direction + 3) % 4.
// When the robot turns right (-1), we update the direction: (direction + 1) % 4.
// When the robot moves forward (k), we iterate k times, updating the coordinates based on the current direction.
// Before each single unit move, we check if the next position is an obstacle. If it is, we stop moving in that direction for this command and move to the next command.
// We use a HashSet to store obstacle coordinates for efficient lookup. The coordinates are stored as a single integer by combining x and y: `y * 60000 + x` (since x, y can range up to 30000).
// We keep track of the maximum squared Euclidean distance from the origin reached at any point.
//
// Time Complexity: O(N + M*K), where N is the number of commands, M is the number of obstacles, and K is the maximum number of steps a robot can take in a single command (which is 9).
// In the worst case, for each command, we might iterate up to 9 times. Checking for obstacles is O(1) on average due to HashSet.
// The dominant factor can be the total number of steps taken by the robot. If the robot moves a total of `TotalSteps` across all commands, the complexity related to movement is O(TotalSteps).
// The obstacle processing is O(M). Thus, the overall time complexity is O(N + M + TotalSteps). Since TotalSteps can be at most N * 9, and N can be up to 10^4, this is approximately O(N + M).
//
// Space Complexity: O(M), where M is the number of obstacles, to store the obstacle locations in a HashSet.
class Solution {
    public int robotSim(int[] commands, int[][] obstacles) {
        // Initialize robot's position and direction
        int x = 0; // current x-coordinate
        int y = 0; // current y-coordinate
        int direction = 0; // 0: North (+y), 1: East (+x), 2: South (-y), 3: West (-x)

        // Define direction vectors for North, East, South, West
        // dx[0]=0, dy[0]=1  -> North
        // dx[1]=1, dy[1]=0  -> East
        // dx[2]=0, dy[2]=-1 -> South
        // dx[3]=-1, dy[3]=0 -> West
        int[] dx = {0, 1, 0, -1};
        int[] dy = {1, 0, -1, 0};

        // Store obstacles in a HashSet for efficient lookup.
        // Each coordinate pair (x, y) is encoded into a single long value.
        // The range of x and y is [-30000, 30000]. A safe offset is 30000.
        // So, we can map x to x + 30000 and y to y + 30000, ensuring positive indices.
        // Then, combine them: (y_offset * range_size) + x_offset.
        // range_size = 2 * 30000 + 1 = 60001. A slightly larger value like 60000 is fine.
        Set<Long> obstacleSet = new HashSet<>();
        for (int[] obstacle : obstacles) {
            // Encode obstacle coordinates into a single long
            obstacleSet.add((long) obstacle[1] * 60000 + obstacle[0]);
        }

        // Variable to store the maximum squared Euclidean distance
        long maxSquaredDistance = 0;

        // Process each command
        for (int command : commands) {
            if (command == -2) { // Turn left
                // Update direction: (direction + 3) % 4 to rotate counter-clockwise
                direction = (direction + 3) % 4;
            } else if (command == -1) { // Turn right
                // Update direction: (direction + 1) % 4 to rotate clockwise
                direction = (direction + 1) % 4;
            } else { // Move forward k units
                int steps = command;
                for (int i = 0; i < steps; i++) {
                    // Calculate the next potential position
                    int nextX = x + dx[direction];
                    int nextY = y + dy[direction];

                    // Check if the next position is an obstacle
                    if (obstacleSet.contains((long) nextY * 60000 + nextX)) {
                        // If it's an obstacle, stop moving in this direction for this command
                        break;
                    }

                    // If not an obstacle, update the robot's position
                    x = nextX;
                    y = nextY;

                    // Update the maximum squared distance
                    // Euclidean distance squared = x^2 + y^2
                    maxSquaredDistance = Math.max(maxSquaredDistance, (long) x * x + (long) y * y);
                }
            }
        }

        // Return the maximum squared Euclidean distance
        return (int) maxSquaredDistance;
    }
}
```

File: solutions/2026-04-06/python/2026-04-06-walking-robot-simulation.py
```python
# Problem Summary:
# Simulate a robot's movement on an infinite grid with obstacles and return the maximum squared Euclidean distance from the origin.
# Link: https://leetcode.com/problems/walking-robot-simulation/
#
# Approach:
# We can simulate the robot's movement step by step. We'll maintain the robot's current position (x, y) and its current direction.
# The directions can be represented as integers: 0 for North (+y), 1 for East (+x), 2 for South (-y), and 3 for West (-x).
# When the robot receives a 'turn left' command (-2), we update the direction by (current_direction - 1) % 4.
# When the robot receives a 'turn right' command (-1), we update the direction by (current_direction + 1) % 4.
# When the robot receives a 'move forward' command (k), we move k steps one by one. For each step, we calculate the next potential position.
# Before moving to the next position, we check if it contains an obstacle. If it does, the robot stays at its current position and the step is effectively skipped.
# If there's no obstacle, we update the robot's current position and calculate the squared Euclidean distance from the origin. We keep track of the maximum squared distance encountered.
# To efficiently check for obstacles, we can store them in a set for O(1) average time lookups.
#
# Time Complexity:
# O(N * K_max + M), where N is the number of commands, K_max is the maximum move command value (9), and M is the number of obstacles.
# In the worst case, each move command of value k might require checking k positions.
# If we consider the total number of unit steps across all commands, it's bounded.
# The total number of steps a robot takes is the sum of all positive commands.
# Let S be the sum of all positive commands. The time complexity is O(S + M) for obstacle lookups. Since the maximum value of a command is 9 and the number of commands is 10^4, S can be up to 9 * 10^4.
# Thus, the time complexity is effectively O(sum(commands) + |obstacles|).
#
# Space Complexity:
# O(M), where M is the number of obstacles, to store the obstacles in a set.

class Solution:
    def robotSim(self, commands: list[int], obstacles: list[list[int]]) -> int:
        # Store obstacles in a set for O(1) average time lookup.
        obstacle_set = set()
        for obs_x, obs_y in obstacles:
            obstacle_set.add((obs_x, obs_y))

        # Robot's current position (x, y)
        x, y = 0, 0
        # Robot's current direction.
        # 0: North (+y)
        # 1: East (+x)
        # 2: South (-y)
        # 3: West (-x)
        direction = 0

        # Maximum squared Euclidean distance reached so far.
        max_sq_dist = 0

        # Define direction vectors for (dx, dy) corresponding to each direction.
        # For example, if direction is 0 (North), dx=0, dy=1.
        dx = [0, 1, 0, -1]
        dy = [1, 0, -1, 0]

        # Iterate through each command.
        for command in commands:
            if command == -1:  # Turn right
                direction = (direction + 1) % 4
            elif command == -2:  # Turn left
                direction = (direction - 1) % 4
            else:  # Move forward k units
                k = command
                # Move one step at a time to check for obstacles at each unit.
                for _ in range(k):
                    # Calculate the next potential position.
                    next_x = x + dx[direction]
                    next_y = y + dy[direction]

                    # Check if the next position is an obstacle.
                    if (next_x, next_y) in obstacle_set:
                        # If it's an obstacle, the robot stays at its current position.
                        # No need to update x, y, just break the inner loop and move to the next command.
                        break
                    else:
                        # If no obstacle, update the robot's position.
                        x = next_x
                        y = next_y
                        # Update the maximum squared distance.
                        max_sq_dist = max(max_sq_dist, x * x + y * y)

        # Return the maximum squared Euclidean distance.
        return max_sq_dist
```